### PR TITLE
Cdd 1566 client worker squash

### PR DESCRIPTION
**Pull request by googley42**

### CODE REVIEW

**Comment by usmanshabbir** [comment]
1) This client worker should be processing things in sequence, currently, a lot of futures are being used which is making the code complicated when it is not required. Once this client is kicked off in an async way, we don't need any async processing inside the worker itself, we are processing each notification one by one either for PUSH or PULL. Simplifying the code will also help to debug and adding new code in future I reckon.

2) The code is not taking the refresh failure into account. we need to stop the loop of processing the notifications when the locking is failed. 

3) When the call goes to PULL, we don't need to fetch the client call back data. I know the model which is being used currently requires that, but we need to change the model and separate the PUSH model from PULL model.

4) when we fail to fetch the collection, the error statement should tell the collection fetch is failed.

5) I think we can extract the PUSH, PULL related code in separate classes, that will make testing this worker easier.

Lets discuss and pair tomorrow, I'll try to get in early tomorrow.

File: app/uk/gov/hmrc/customs/notification/services/ClientWorkerImpl.scala [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                      ) extends ClientWorker {
+
+  private val extendLockDuration =  org.joda.time.Duration.millis(config.pushNotificationConfig.lockRefreshDurationInMilliseconds)
+  private val refreshDuration = Duration(config.pushNotificationConfig.lockRefreshDurationInMilliseconds, TimeUnit.MILLISECONDS)
+
+  //TODO: should we pass in HeaderCarrier as an implicit parameter for logging?
+  override def processNotificationsFor(csid: ClientSubscriptionId, lockOwnerId: LockOwnerId): Future[Unit] = {
+    //implicit HeaderCarrier required for ApiSubscriptionFieldsConnector
+    //however looking at api-subscription-fields service I do not think it is required so keep new HeaderCarrier() for now
+    implicit val hc = HeaderCarrier()
```
Comment (usmanshabbir) [inline app/uk/gov/hmrc/customs/notification/services/ClientWorkerImpl.scala]: it should receive the expiry time from the dispatcher which locked this record. The refresher should refresh it when that time runs out and from that point onwards, it should be running after refreshDuration. The hc can be defined on top of this class and can be used wherever it is required.

File: app/uk/gov/hmrc/customs/notification/services/ClientWorkerImpl.scala [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+    logger.info(s"About to process notifications")
+
+    (for {
+      clientNotifications <- repo.fetch(csid)
+      _ <- sequence(clientNotifications)(pushClientNotification)
+    } yield ())
+    .recover {
+      case e: Throwable =>
+        logger.error("Error pushing notification")
```
Comment (usmanshabbir) [inline app/uk/gov/hmrc/customs/notification/services/ClientWorkerImpl.scala]: here the error could be different then pushing, if we are failed to fetch the notifications then we should just exit from here and not send it to Pull.

File: app/uk/gov/hmrc/customs/notification/services/ClientWorkerImpl.scala [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      case e: Exception =>
+        val msg = e.getMessage
+        logger.error(msg) //TODO: extend logging API so that we can log an error on a throwable
+    }
+  }
+
+
+  protected def process(csid: ClientSubscriptionId)(implicit hc: HeaderCarrier): Future[Unit] = {
+
+    logger.info(s"About to process notifications")
```
Comment (rashadmughal) [inline app/uk/gov/hmrc/customs/notification/services/ClientWorkerImpl.scala]: what is the guidance regarding info messages? @usmanshabbir One per successful 'request' (or poll?)

**Pull request closed without merging**